From 5c49cb8f884ecf227e44159a5a4931d2c0814d5c Mon Sep 17 00:00:00 2001
Message-Id: <5c49cb8f884ecf227e44159a5a4931d2c0814d5c.1434670190.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Mon, 25 May 2015 16:15:21 +0530
Subject: [PATCH 1/6] FOR_UPSTREAM [VPG]: drm/i915: Add DPI AND DBI FIFO empty
 checks

DBI DPI FIFO empty bits in the MIPI_FIFO_STAT register needs
to be checked to make sure the FIFOs are empty during DSI
disable.

Issue: GMINL-10288
Change-Id: I98a90d14cb3568b1ece7c4681b190e37e53ba2aa
Signed-off-by: Yogesh Mohan Marimuthu <yogesh.mohan.marimuthu@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/intel_dsi.c     |  2 ++
 drivers/gpu/drm/i915/intel_dsi.h     |  1 +
 drivers/gpu/drm/i915/intel_dsi_cmd.c | 28 ++++++++++++++++++++++++++++
 3 files changed, 31 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_dsi.c b/drivers/gpu/drm/i915/intel_dsi.c
index 3340bc8..d9eaaac 100644
--- a/drivers/gpu/drm/i915/intel_dsi.c
+++ b/drivers/gpu/drm/i915/intel_dsi.c
@@ -708,6 +708,8 @@ static void intel_dsi_post_disable(struct intel_encoder *encoder)
 
 	DRM_DEBUG_KMS("\n");
 
+	wait_for_dpi_dbi_fifo_empty(intel_dsi);
+
 	intel_dsi_disable(encoder);
 
 	intel_dsi_clear_device_ready(encoder);
diff --git a/drivers/gpu/drm/i915/intel_dsi.h b/drivers/gpu/drm/i915/intel_dsi.h
index ffa0f3d..96bdac2 100644
--- a/drivers/gpu/drm/i915/intel_dsi.h
+++ b/drivers/gpu/drm/i915/intel_dsi.h
@@ -543,6 +543,7 @@ static inline bool is_cmd_mode(struct intel_dsi *intel_dsi)
 extern void vlv_enable_dsi_pll(struct intel_encoder *encoder);
 extern void vlv_disable_dsi_pll(struct intel_encoder *encoder);
 extern void wait_for_dsi_fifo_empty(struct intel_dsi *intel_dsi);
+extern void wait_for_dpi_dbi_fifo_empty(struct intel_dsi *intel_dsi);
 extern u32 vlv_get_dsi_pclk(struct intel_encoder *encoder, int pipe_bpp);
 extern struct intel_dsi_dev_ops vbt_generic_dsi_display_ops;
 extern struct intel_dsi *intel_attached_dsi(struct drm_connector *connector);
diff --git a/drivers/gpu/drm/i915/intel_dsi_cmd.c b/drivers/gpu/drm/i915/intel_dsi_cmd.c
index d2c7787..63d642b 100644
--- a/drivers/gpu/drm/i915/intel_dsi_cmd.c
+++ b/drivers/gpu/drm/i915/intel_dsi_cmd.c
@@ -495,6 +495,34 @@ int dpi_send_cmd(struct intel_dsi *intel_dsi, u32 cmd, bool hs)
 	return 0;
 }
 
+void wait_for_dpi_dbi_fifo_empty(struct intel_dsi *intel_dsi)
+{
+	struct drm_encoder *encoder = &intel_dsi->base.base;
+	struct drm_device *dev = encoder->dev;
+	struct drm_i915_private *dev_priv = dev->dev_private;
+	struct intel_crtc *intel_crtc = to_intel_crtc(encoder->crtc);
+	enum pipe pipe = intel_crtc->pipe;
+	u32 mask;
+	int count = 1;
+
+	mask = DPI_FIFO_EMPTY | DBI_FIFO_EMPTY;
+
+	if (intel_dsi->dual_link) {
+		count = 2;
+		pipe = PIPE_A;
+	}
+
+	do {
+		if (wait_for((I915_READ(MIPI_GEN_FIFO_STAT(pipe)) & mask)
+								== mask, 100))
+			DRM_ERROR("DPI/DBI FIFO empty wait timed out\n");
+
+		/* For Port C for dual link */
+		if (intel_dsi->dual_link)
+			pipe = PIPE_B;
+	} while (--count > 0);
+}
+
 void wait_for_dsi_fifo_empty(struct intel_dsi *intel_dsi)
 {
 	struct drm_encoder *encoder = &intel_dsi->base.base;
-- 
1.9.1

